modules/placement-driver/src/main/java/org/apache/ignite/internal/placementdriver/LeaseUpdater.java

```diff
@@ -0,0 +1,242 @@
+     * Runnable to update lease in Metastorage.
+     */
+    private class Updater implements Runnable {
+        @Override
+        public void run() {
+            while (!updaterTread.isInterrupted()) {
+                for (Map.Entry<ReplicationGroupId, Set<Assignment>> entry : assignmentsTracker.assignments().entrySet()) {
+                    ReplicationGroupId grpId = entry.getKey();
+
+                    Lease lease = leaseTracker.getLease(grpId);
+
+                    HybridTimestamp now = clock.now();
+
+                    // Nothing holds the lease.
+                    if (lease.getLeaseExpirationTime() == null
+                            // The lease is near to expiration.
+                            || now.getPhysical() > (lease.getLeaseExpirationTime().getPhysical() - LEASE_PERIOD / 2)) {
+                        var leaseKey = ByteArray.fromString(PLACEMENTDRIVER_PREFIX + grpId);
+
+                        var newTs = new HybridTimestamp(now.getPhysical() + LEASE_PERIOD, 0);
+
+                        ClusterNode candidate = nextLeaseHolder(entry.getValue());
+
+                        // Now nothing is holding the lease, and new lease is being granted.
+                        if (lease.getLeaseholder() == null && candidate != null
+                                // The lease is prolongation
+                                || lease.getLeaseholder() != null && lease.getLeaseholder().equals(candidate)
+                                // New lease is being granted, and the previous lease is expired.
+                                || candidate != null && (lease.getLeaseExpirationTime() == null
+                                || compareWithClockSkew(now, lease.getLeaseExpirationTime()) > 0)) {
+                            byte[] leaseRaw = ByteUtils.toBytes(lease);
+
+                            Lease renewedLease = new Lease(candidate, newTs);
+
+                            msManager.invoke(
+                                    or(notExists(leaseKey), value(leaseKey).eq(leaseRaw)),
+                                    put(leaseKey, ByteUtils.toBytes(renewedLease)),
+                                    noop()
+                            );
+                        }
+                    }
+                }
+
+                try {
+                    Thread.sleep(UPDATE_LEASE_MS);
+                } catch (InterruptedException e) {
+                    LOG.warn("Lease updater is interrupted");
+                }
+            }
+        }
+    }
+
+    /**
+     * Compares two timestamps with the clock skew.
+     * TODO: IGNITE-18978 Method to comparison timestamps with clock skew.
+     *
+     * @param ts1 First timestamp.
+     * @param ts2 Second timestamp.
+     * @return Result of comparison.
+     */
+    private static int compareWithClockSkew(HybridTimestamp ts1, HybridTimestamp ts2) {
+        if (ts1.getPhysical() - CLOCK_SKEW < ts2.getPhysical() && ts1.getPhysical() + CLOCK_SKEW < ts2.getPhysical()) {
```

Comment:
This condition is incorrect: if the second part is true, the first should be also true. Perhaps, you mean `>` in the second part.
But actually, we can't claim the equality of two timestamps if this method returns `0`, in fact it means that we don't know did one happen before or after another or not. This this should be definitely mentioned because its counter-intuitive and contradicts with traditional comparator pattern.

Reply:
Yes, here is a mistake to use `&&` instead of `||`.
About other part of your comment. You told about Comparable contract, but here is a method (not implementation of method compare). If you think the name `compareWithClockSkew` is not applicable and misled, I can rename it.